### ISSUE TITLE
udevadm settle taking more than 120s

### ISSUE THREAD

**Issue by yizhanglinux** [systemd/systemd#23333]
**systemd version the issue has been seen with**
 > ` systemd-udev-251~rc2-1.fc37.aarch64 systemd-251~rc2-1.fc37.aarch64`

**Used distribution**
 > Fedora-Rawhide

**Linux kernel version used** (`uname -a`)

 > `5.18.0-0.rc5.20220506gitfe27d189e3f42e3.44.fc37.aarch64`

**CPU architecture issue was seen on**
 > aarch64

**Expected behaviour you didn't see**
 >  The cmd should be finished soon

**Unexpected behaviour you saw**
 > Everytime executing the cmd take more than 120s

**Steps to reproduce the problem**
 >  
$ modprobe null-blk
$ time udevadm settle 

real	2m0.061s
user	0m0.174s
sys	0m6.809s

**Additional program output to the terminal or log subsystem illustrating the issue**

```text
…
```

**Comment by poettering** [comment]
most likely some udev jobs are still running?

please provide debug logging when this happens, otherwise this isn't really actionable

**Comment by yizhanglinux** [comment]
Here is the executing and journalctl log:

```
+ echo 'START: Tue May 10 10:37:04 PM EDT 2022'
+ rmmod null-blk
+ modprobe null-blk
+ udevadm settle

real	2m0.058s
user	0m0.221s
sys	0m8.414s
+ echo 'END: Tue May 10 10:39:04 PM EDT 2022'
+ sleep 120
+ journalctl --no-hostname -b -u systemd-udevd.service
```

# journalctl --no-hostname -b -u systemd-udevd.service
```
May 10 22:37:04 systemd-udevd[1142]: 251:0: Device is queued (SEQNUM=7881, ACTION=remove)
May 10 22:37:04 systemd-udevd[1142]: Validate module index
May 10 22:37:04 systemd-udevd[1142]: Check if link configuration needs reloading.
May 10 22:37:04 systemd-udevd[1142]: 251:0: Device ready for processing (SEQNUM=7881, ACTION=remove)
May 10 22:37:04 systemd-udevd[1142]: Successfully forked off 'n/a' as PID 2775.
May 10 22:37:04 systemd-udevd[1142]: 251:0: Worker [2775] is forked for processing SEQNUM=7881.
May 10 22:37:04 systemd-udevd[1142]: nullb0: Device is queued (SEQNUM=7882, ACTION=remove)
May 10 22:37:04 systemd-udevd[1142]: nullb0: Device ready for processing (SEQNUM=7882, ACTION=remove)
May 10 22:37:04 systemd-udevd[2775]: 251:0: Processing device (SEQNUM=7881, ACTION=remove)
May 10 22:37:04 systemd-udevd[1142]: Successfully forked off 'n/a' as PID 2776.
May 10 22:37:04 systemd-udevd[1142]: nullb0: Worker [2776] is forked for processing SEQNUM=7882.
May 10 22:37:04 systemd-udevd[2775]: 251:0: Device processed (SEQNUM=7881, ACTION=remove)
May 10 22:37:04 systemd-udevd[2775]: 251:0: sd-device-monitor: Passed 144 byte to netlink monitor
May 10 22:37:04 systemd-udevd[2776]: nullb0: Processing device (SEQNUM=7882, ACTION=remove)
May 10 22:37:04 systemd-udevd[2776]: nullb0: Failed to get watch handle, ignoring: No such file or directory
May 10 22:37:04 systemd-udevd[2776]: nullb0: Device processed (SEQNUM=7882, ACTION=remove)
May 10 22:37:04 systemd-udevd[2776]: nullb0: sd-device-monitor: Passed 278 byte to netlink monitor
May 10 22:37:04 systemd-udevd[1142]: No events are queued, removing /run/udev/queue.
May 10 22:37:04 systemd-udevd[1142]: null_blk: Device is queued (SEQNUM=7883, ACTION=remove)
May 10 22:37:04 systemd-udevd[1142]: null_blk: Device ready for processing (SEQNUM=7883, ACTION=remove)
May 10 22:37:04 systemd-udevd[1142]: null_blk: sd-device-monitor: Passed 137 byte to netlink monitor
May 10 22:37:04 systemd-udevd[2775]: null_blk: Processing device (SEQNUM=7883, ACTION=remove)
May 10 22:37:04 systemd-udevd[2775]: null_blk: Device processed (SEQNUM=7883, ACTION=remove)
May 10 22:37:04 systemd-udevd[2775]: null_blk: sd-device-monitor: Passed 137 byte to netlink monitor
May 10 22:37:04 systemd-udevd[1142]: No events are queued, removing /run/udev/queue.
May 10 22:37:04 systemd-udevd[1142]: 251:0: Device is queued (SEQNUM=7884, ACTION=add)
May 10 22:37:04 systemd-udevd[1142]: 251:0: Device ready for processing (SEQNUM=7884, ACTION=add)
May 10 22:37:04 systemd-udevd[1142]: 251:0: sd-device-monitor: Passed 141 byte to netlink monitor
May 10 22:37:04 systemd-udevd[2775]: 251:0: Processing device (SEQNUM=7884, ACTION=add)
May 10 22:37:04 systemd-udevd[1142]: nullb0: Device is queued (SEQNUM=7885, ACTION=add)
May 10 22:37:04 systemd-udevd[1142]: nullb0: Device ready for processing (SEQNUM=7885, ACTION=add)
May 10 22:37:04 systemd-udevd[1142]: nullb0: sd-device-monitor: Passed 208 byte to netlink monitor
May 10 22:37:04 systemd-udevd[2776]: nullb0: Processing device (SEQNUM=7885, ACTION=add)
May 10 22:37:04 systemd-udevd[2776]: nullb0: Failed to get watch handle, ignoring: No such file or directory
May 10 22:37:04 systemd-udevd[2776]: nullb0: /usr/lib/udev/rules.d/50-udev-default.rules:74 GROUP 6
May 10 22:37:04 systemd-udevd[2775]: 251:0: /usr/lib/udev/rules.d/99-nfs.rules:1 Running PROGRAM '/usr/libexec/nfsrahead 251:0'
May 10 22:37:04 systemd-udevd[2775]: 251:0: Starting '/usr/libexec/nfsrahead 251:0'
May 10 22:37:04 systemd-udevd[2776]: nullb0: Setting permissions /dev/nullb0, uid=0, gid=6, mode=0660
May 10 22:37:04 systemd-udevd[1142]: null_blk: Device is queued (SEQNUM=7886, ACTION=add)
May 10 22:37:04 systemd-udevd[1142]: null_blk: Device ready for processing (SEQNUM=7886, ACTION=add)
May 10 22:37:04 systemd-udevd[2776]: nullb0: Handling device node '/dev/nullb0', devnum=b251:0
May 10 22:37:04 systemd-udevd[2775]: Successfully forked off '(spawn)' as PID 2778.
May 10 22:37:04 systemd-udevd[2776]: nullb0: sd-device: Created db file '/run/udev/data/b251:0' for '/devices/virtual/block/nullb0'
May 10 22:37:04 systemd-udevd[2776]: nullb0: Failed to get watch handle, ignoring: No such file or directory
May 10 22:37:04 systemd-udevd[2776]: nullb0: Device processed (SEQNUM=7885, ACTION=add)
May 10 22:37:04 systemd-udevd[2776]: nullb0: sd-device-monitor: Passed 275 byte to netlink monitor
May 10 22:37:04 systemd-udevd[1142]: Successfully forked off 'n/a' as PID 2779.
May 10 22:37:04 systemd-udevd[1142]: null_blk: Worker [2779] is forked for processing SEQNUM=7886.
May 10 22:37:04 systemd-udevd[2779]: null_blk: Processing device (SEQNUM=7886, ACTION=add)
May 10 22:37:04 systemd-udevd[2779]: null_blk: Device processed (SEQNUM=7886, ACTION=add)
May 10 22:37:04 systemd-udevd[2779]: null_blk: sd-device-monitor: Passed 134 byte to netlink monitor
May 10 22:37:04 systemd-udevd[1142]: Received udev control message (PING)
May 10 22:38:04 systemd-udevd[1142]: 251:0: Worker [2775] processing SEQNUM=7884 is taking a long time
May 10 22:38:04 systemd-udevd[2775]: 251:0: Spawned process '/usr/libexec/nfsrahead 251:0' [2778] is taking longer than 59s to complete
May 10 22:38:04 systemd-udevd[1142]: Validate module index
May 10 22:38:04 systemd-udevd[1142]: Check if link configuration needs reloading.
May 10 22:40:04 systemd-udevd[2775]: 251:0: Spawned process '/usr/libexec/nfsrahead 251:0' [2778] timed out after 2min 59s, killing
May 10 22:40:04 systemd-udevd[1142]: 251:0: Worker [2775] processing SEQNUM=7884 killed
May 10 22:40:04 systemd-udevd[1142]: Validate module index
May 10 22:40:04 systemd-udevd[1142]: Check if link configuration needs reloading.
May 10 22:40:04 systemd-udevd[1142]: 251:0: Worker [2775] terminated by signal 9 (KILL).
May 10 22:40:04 systemd-udevd[1142]: 251:0: sd-device-monitor: Passed 212 byte to netlink monitor
May 10 22:40:04 systemd-udevd[1142]: No events are queued, removing /run/udev/queue.
May 10 22:40:08 systemd-udevd[1142]: Cleanup idle workers
May 10 22:40:08 systemd-udevd[2779]: Unload module index
May 10 22:40:08 systemd-udevd[2776]: Unload module index
May 10 22:40:08 systemd-udevd[2776]: Unloaded link configuration context.
May 10 22:40:08 systemd-udevd[2779]: Unloaded link configuration context.
May 10 22:40:08 systemd-udevd[1142]: Worker [2779] exited.
May 10 22:40:08 systemd-udevd[1142]: Worker [2776] exited.
```

**Comment by yuwata** [comment]
As you can see, `nfsrahead` takes a long time. What package provides 99-nfs.rules? Please adjust it. 
```
May 10 22:37:04 systemd-udevd[2775]: 251:0: /usr/lib/udev/rules.d/99-nfs.rules:1 Running PROGRAM '/usr/libexec/nfsrahead 251:0'
May 10 22:37:04 systemd-udevd[2775]: 251:0: Starting '/usr/libexec/nfsrahead 251:0'
May 10 22:38:04 systemd-udevd[1142]: 251:0: Worker [2775] processing SEQNUM=7884 is taking a long time
May 10 22:38:04 systemd-udevd[2775]: 251:0: Spawned process '/usr/libexec/nfsrahead 251:0' [2778] is taking longer than 59s to complete
May 10 22:40:04 systemd-udevd[2775]: 251:0: Spawned process '/usr/libexec/nfsrahead 251:0' [2778] timed out after 2min 59s, killing
May 10 22:40:04 systemd-udevd[1142]: 251:0: Worker [2775] processing SEQNUM=7884 killed
May 10 22:40:04 systemd-udevd[1142]: 251:0: Worker [2775] terminated by signal 9 (KILL).
May 10 22:40:04 systemd-udevd[1142]: 251:0: sd-device-monitor: Passed 212 byte to netlink monitor
```

This is not a bug in systemd-udevd. Closing.

**Comment by yuwata** [comment]
This seems already fixed by nfs-utils-2.6.1-2.rc4. See https://bugzilla.redhat.com/show_bug.cgi?id=2077585.

**Comment by yizhanglinux** [comment]
> Spawned process '/usr/libexec/nfsrahead 251:0' [2778] is taking longer than 59s to complete

Actually, I'm using the latest version nfs-utils-2.6.1-2.rc4.fc37.aarch64, will open another BZ for Fedora to track it, thanks for your update.

**Comment by yizhanglinux** [comment]
> This seems already fixed by nfs-utils-2.6.1-2.rc4. See https://bugzilla.redhat.com/show_bug.cgi?id=2077585.

Actually, I'm using the latest version nfs-utils-2.6.1-2.rc4.fc37.aarch64, will open another BZ for Fedora to track it, thanks for your update.
